android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationEngineState.kt

```diff
@@ -19,3 +19,81 @@ data class TranslationEngineState(
 )
+
+/**
+ * Determines the best initial "to" language based on the translation state and user preferred
+ * languages.
+ *
+ * @param candidateLanguages The language options available to select as a final initial value.
+ * @return The best determined "to" language or null if a determination cannot be made.
+ */
+fun TranslationEngineState.initialToLanguage(candidateLanguages: List<Language>?): Language? {
+    return determineInitialDropdownValue(
+        translationEngineState = this,
+        isFromLanguage = false,
+        candidateLanguages = candidateLanguages,
+    )
+}
+
+/**
+ * Determines the best initial "from" language based on the translation state and page state.
+ *
+ * @param candidateLanguages The language options available to select as a final initial value.
+ * @return The best determined "from" language or null if a determination cannot be made.
+ */
+fun TranslationEngineState.initialFromLanguage(candidateLanguages: List<Language>?): Language? {
+    return determineInitialDropdownValue(
+        translationEngineState = this,
+        isFromLanguage = true,
+        candidateLanguages = candidateLanguages,
+    )
+}
+
+/**
+ * Helper function to calculate what the initial expected "to" and "from" language values. Will
+ * return null if a determination cannot be made.
+ *
+ * Goal is to first identify if the page was translated, if yes, then present those values as
+ * initial. If the page was not translated, then the page language and user preferred languages
+ * should be prioritized. If the determined language is not a selection, then return null.
+ *
+ * @param translationEngineState Translation engine state to base determination values from.
+ * @param isFromLanguage True when calculating the initial "from" value.
+ * False when calculating the initial "to" value.
+ * @param candidateLanguages The list of potential languages that the candidate language may be
+ * selected from.
+ *
+ * @return The candidate language for the initial value.
+ */
+private fun determineInitialDropdownValue(
+    translationEngineState: TranslationEngineState,
+    isFromLanguage: Boolean,
+    candidateLanguages: List<Language>?,
+): Language? {
+    var code: String?
+
+    // If these values are not null, then the page was translated:
+    code = if (isFromLanguage) {
+        translationEngineState.requestedTranslationPair?.fromLanguage
+    } else {
+        translationEngineState.requestedTranslationPair?.toLanguage
+    }
+
+    // If the page wasn't translated, then use these values:
+    if (code == null) {
+        code = if (isFromLanguage) {
+            translationEngineState.detectedLanguages?.documentLangTag
+        } else {
+            translationEngineState.detectedLanguages?.userPreferredLangTag
+        }
+    }
+
+    // Process code to a known language in our candidate language lists:
+    fun languageCodeToLanguage(bcpCode: String?, list: List<Language>?): Language? {
```

Comment:
Just FYI, nesting functions inside each other is not a typical Kotlin structure. There are some examples of it in the codebase but I would ignore those 😅

Reply:
Thanks for mentioning, good to know, updated!